### PR TITLE
Seasonal abilities

### PR DESCRIPTION
**Pull request by Melody-Ruth**
Contrary to the branch name, this PR contains all four seasonal abilities.
Pressing right click launches a projectile from the player in the direction that the cursor is pointing at (similar to bullet targeting), with a cooldown for how often players can launch a new projectile. The projectile will go forward until it hits the environment or another player, then it will explode (currently just disappear, later we should add particles). When it explodes there's an effect on all players within a radius, with a stronger/longer effect for closer players. The effect is based on the season of the player who created the projectile: spring is healing (up to a cap), summer is confuse controls (swap w/s and a/d), autumn is knockback, and winter is slowness.

### CODE REVIEW

File: server/src/bge/System.cpp [hunk 1]
```diff
@@ -423,26 +441,150 @@ namespace bge {
...
+                    else if (projData.type == SUMMER) {
+                        if (distFromExplosion < PROJ_EXPLOSION_RADIUS) {
+                            statusEffects.swappedControlsTicksLeft = (PROJ_EXPLOSION_RADIUS - distFromExplosion) * (PROJ_EXPLOSION_RADIUS - distFromExplosion) * MAX_PROJ_EFFECT_LENGTH / (PROJ_EXPLOSION_RADIUS * PROJ_EXPLOSION_RADIUS);
+                        }
+                    }
+                    else if (projData.type == AUTUMN) {
+                        if (distFromExplosion < PROJ_EXPLOSION_RADIUS) {
+                            glm::vec3 launchDir = glm::normalize(playerPos.position - pos.position);
+                            float launchSeverity = (PROJ_EXPLOSION_RADIUS - distFromExplosion) * (PROJ_EXPLOSION_RADIUS - distFromExplosion) * MAX_LAUNCH_SEVERITY / (PROJ_EXPLOSION_RADIUS * PROJ_EXPLOSION_RADIUS);
+                            playerVel.velocity += launchSeverity * launchDir;
```
Comment (Alanine42) [inline server/src/bge/System.cpp:578]: i tried out winter, spring and summer's abilities, they're so cool:) !

when I throw the autumn ball, I somehow get teleported immediately into the void.... my velocity is printed (nan,nan,nan)

what might happen:
- the autumn projectile collides immediately with its thrower
- thus `playerPos.position - pos.position` is a zero vector
- normalizing a zero vector will gg

fix: just launch the projectile from outside the player (I think you guys already did this but didnt update projPos.position )

**Review by Alanine42 — APPROVED** [review]

File: server/src/bge/System.cpp (hunk 1 above)
Comment (Melody-Ruth) [reply server/src/bge/System.cpp:579]: What exactly did you do? I can't reproduce the bug with autumn you're describing (I haven't pulled your latest commit).

File: client/include/ClientGame.h [hunk 2]
```diff
@@ -81,6 +81,7 @@ class ClientGame
     bool requestRightward = false;
     bool requestJump = false;
     bool requestThrowEgg = false;
+    bool requestSeasonAbility = false;
```
Comment (benjxia) [inline client/include/ClientGame.h:84]: Is this ever used?

**Review by benjxia — APPROVED** [review]
overall looks good, will still need to integrate with new particle vfx later though

File: client/include/ClientGame.h (hunk 2 above)
Comment (Melody-Ruth) [reply client/include/ClientGame.h:84]: Nope, good catch. I missed that while merging.

**Review by realmatthewpeng — APPROVED** [review]
Looks good to me. Only thing is maybe we should set a min & max launchSeverity. Otherwise (check the video in discord) the player can get launched super high or not get pushed back that much.

**Comment by Melody-Ruth** [comment]
> Looks good to me. Only thing is maybe we should set a min & max launchSeverity. Otherwise (check the video in discord) the player can get launched super high or not get pushed back that much.

There's already a max and min launch severity, so I'll turn the max launch severity down a bit. (min launch severity is just the severity gradually reaching 0 as you reach the limit of the explosion radius)

**Review by Metarus — APPROVED** [review]
lgtm

**Pull request merged**